File: and-bible/app/src/main/java/net/bible/android/view/activity/base/ActivityBase.java
package net.bible.android.view.activity.base;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.KeyEvent;
import android.view.View;
import android.view.WindowManager;

import net.bible.android.BibleApplication;
import net.bible.android.view.activity.ActivityComponent;
import net.bible.android.view.activity.DaggerActivityComponent;
import net.bible.android.view.util.locale.LocaleHelper;
import net.bible.android.view.activity.navigation.History;
import net.bible.android.view.activity.page.MainBibleActivity;
import net.bible.android.view.util.UiUtils;
import net.bible.service.common.CommonUtils;
import net.bible.service.device.ScreenSettings;
import net.bible.service.history.HistoryTraversal;
import net.bible.service.history.HistoryTraversalFactory;
import net.bible.service.sword.SwordDocumentFacade;

import javax.inject.Inject;

/** Base class for activities
 * 
 * @author Martin Denham [mjdenham at gmail dot com]
 * @see gnu.lgpl.License for license details.<br>
 *      The copyright to this program is held by it's author.
 */
public class ActivityBase extends AppCompatActivity implements AndBibleActivity {

	// standard request code for startActivityForResult
	public static final int STD_REQUEST_CODE = 1;
	
	// Special result that requests all activities to exit until the main/top Activity is reached
    public static final int RESULT_RETURN_TO_TOP           = 900;

	private SharedActivityState sharedActivityState = SharedActivityState.getInstance();
	
	private boolean isScreenOn = true;

	// some screens are highly customised and the theme looks odd if it changes
	private boolean allowThemeChange = true;
	
	private View mContentView;

	private HistoryTraversal historyTraversal;

	private boolean integrateWithHistoryManagerInitialValue;

	private SwordDocumentFacade swordDocumentFacade;
	
	private static final String TAG = "ActivityBase";
	
	/** Called when the activity is first created. */
    @SuppressLint("MissingSuperCall")
	@Override
    public void onCreate(Bundle savedInstanceState) {
    	this.onCreate(savedInstanceState, false);
    }

    /** Called when the activity is first created. */
    public void onCreate(Bundle savedInstanceState, boolean integrateWithHistoryManager) {
    	if (allowThemeChange) {
    		UiUtils.applyTheme(this);
    	}

		super.onCreate(savedInstanceState);
    	
        Log.i(getLocalClassName(), "onCreate:"+this);

		this.integrateWithHistoryManagerInitialValue = integrateWithHistoryManager;

        // Register current activity in onCreate and onResume
        CurrentActivityHolder.getInstance().setCurrentActivity(this);

        // fix for null context class loader (http://code.google.com/p/android/issues/detail?id=5697)
        // this affected jsword dynamic classloading
        Thread.currentThread().setContextClassLoader(getClass().getClassLoader());
		
        setFullScreen(isFullScreen());

		// if locale is overridden then have to force title to be translated here
		LocaleHelper.translateTitle(this);
    }

	protected ActivityComponent buildActivityComponent() {
		return DaggerActivityComponent.builder()
				.applicationComponent(BibleApplication.getApplication().getApplicationComponent())
				.build();
	}

    @Override
	public void startActivity(Intent intent) {
    	historyTraversal.beforeStartActivity();
    	
		super.startActivity(intent);
	}
	@Override
	public void startActivityForResult(Intent intent, int requestCode) {
    	historyTraversal.beforeStartActivity();

    	super.startActivityForResult(intent, requestCode);
	}

	/**
	 * Override locale.  If user has selected a different ui language to the devices default language
	 */
	@Override
	protected void attachBaseContext(Context newBase) {
		super.attachBaseContext(LocaleHelper.onAttach(newBase));
	}

	/**	This will be called automatically for you on 2.0 or later
	 */
	@Override
	public void onBackPressed() {
		if (!historyTraversal.goBack()) {
			super.onBackPressed();
		}
	}
	
	/**
	 * Change fullscreen state for this and all future activities
	 */
    public void toggleFullScreen() {
    	sharedActivityState.toggleFullScreen();
    	setFullScreen(sharedActivityState.isFullScreen());
    }
	
    /**
     * Are all activities currently in full screen mode
     */
	public boolean isFullScreen() {
		return sharedActivityState.isFullScreen();
	}
	
	private void setFullScreen(boolean isFullScreen) {
    	if (!isFullScreen) {
    		Log.d(TAG, "NOT Fullscreen");
    		// http://stackoverflow.com/questions/991764/hiding-title-in-a-fullscreen-mode
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_FORCE_NOT_FULLSCREEN);
            getWindow().clearFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
			setLightsOutMode(false);
    	} else {
    		Log.d(TAG, "Fullscreen");
            getWindow().addFlags(WindowManager.LayoutParams.FLAG_FULLSCREEN);
            getWindow().clearFlags(WindowManager.LayoutParams.FLAG_FORCE_NOT_FULLSCREEN);
			setLightsOutMode(true);
    	}
	}
	
	@SuppressLint("NewApi")
	private void setLightsOutMode(boolean isLightsOut) {
		if (CommonUtils.isHoneycombPlus() && mContentView!=null) {
	    	if (isLightsOut) {
				mContentView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_LOW_PROFILE);
	    	} else {
				mContentView.setSystemUiVisibility(View.SYSTEM_UI_FLAG_VISIBLE);
	    	}
		}
	}

	/** called by Android 2.0 +
	 */
	@Override
	public boolean onKeyLongPress(int keyCode, KeyEvent event) {
		// ignore long press on search because it causes errors
		if (keyCode == KeyEvent.KEYCODE_SEARCH) {
	    	// ignore
			return true;
		}
		
		//TODO make Long press Back work for screens other than main window e.g. does not work from search screen because wrong window is displayed
	    if (keyCode == KeyEvent.KEYCODE_BACK && this instanceof MainBibleActivity) {
			Log.d(TAG, "Back Long");
	        // a long press of the back key. do our work, returning true to consume it.  by returning true, the framework knows an action has
	        // been performed on the long press, so will set the cancelled flag for the following up event.
	    	Intent intent = new Intent(this, History.class);
	    	startActivityForResult(intent, 1);
	        return true;
	    }
	    
		//TODO make Long press back - currently the History screen does not show the correct screen after item selection if not called from main window 
	    if (keyCode == KeyEvent.KEYCODE_BACK) {
	    	// ignore
	    	return true;
	    }

	    return super.onKeyLongPress(keyCode, event);
	}

	@Override
	public boolean isIntegrateWithHistoryManager() {
		return historyTraversal.isIntegrateWithHistoryManager();
	}
	@Override
	public void setIntegrateWithHistoryManager(boolean integrateWithHistoryManager) {
		historyTraversal.setIntegrateWithHistoryManager(integrateWithHistoryManager);
	}
	
    /** allow activity to enhance intent to correctly restore state */
	public Intent getIntentForHistoryList() {
		return getIntent();
	}

	public void showErrorMsg(int msgResId) {
		Dialogs.getInstance().showErrorMsg(msgResId);
	}

    protected void showHourglass() {
    	Dialogs.getInstance().showHourglass();
    }
    protected void dismissHourglass() {
    	Dialogs.getInstance().dismissHourglass();
    }

    protected void returnErrorToPreviousScreen() {
    	// just pass control back to the previous screen
    	Intent resultIntent = new Intent(this, this.getClass());
    	setResult(Activity.RESULT_CANCELED, resultIntent);
    	finish();    
    }
    protected void returnToPreviousScreen() {
    	// just pass control back to the previous screen
    	Intent resultIntent = new Intent(this, this.getClass());
    	setResult(Activity.RESULT_OK, resultIntent);
    	finish();    
    }
    
    protected void returnToTop() {
    	// just pass control back to the previous screen
    	Intent resultIntent = new Intent(this, this.getClass());
    	setResult(RESULT_RETURN_TO_TOP, resultIntent);
    	finish();    
    }
    
	@Override
	protected void onResume() {
		super.onResume();
        Log.i(getLocalClassName(), "onResume:"+this);
        CurrentActivityHolder.getInstance().setCurrentActivity(this);
        
        //allow action to be called on screen being turned on
		if (!isScreenOn && ScreenSettings.isScreenOn()) {
			onScreenTurnedOn();
		}
	}

	@Override
	protected void onPause() {
		super.onPause();
        Log.i(getLocalClassName(), "onPause:"+this);
		if (isScreenOn && !ScreenSettings.isScreenOn()) {
			onScreenTurnedOff();
		}
	}
	
	protected void onScreenTurnedOff() {
		Log.d(TAG, "Window turned off");
		isScreenOn = false;
	}

	protected void onScreenTurnedOn() {
		Log.d(TAG, "Window turned on");
		isScreenOn = true;
	}

	@Override
	protected void onRestart() {
		super.onRestart();
        Log.i(getLocalClassName(), "onRestart:"+this);
	}

	@Override
	protected void onStart() {
		super.onStart();
        Log.i(getLocalClassName(), "onStart:"+this);
	}


	@Override
	protected void onStop() {
		super.onStop();
        Log.i(getLocalClassName(), "onStop:"+this);
        // screen can still be considered as current screen if put on stand-by
        // removing this if causes speech to stop when screen is put on stand-by
        if (isScreenOn) {
	        // call this onStop, although it is not guaranteed to be called, to ensure an overlap between dereg and reg of current activity, otherwise AppToBackground is fired mistakenly
	        CurrentActivityHolder.getInstance().iAmNoLongerCurrent(this);
        }
	}

	public void setAllowThemeChange(boolean allowThemeChange) {
		this.allowThemeChange = allowThemeChange;
	}
	
	/** custom title bar code to add the FEATURE_CUSTOM_TITLE just before setContentView
	 * and set the new titlebar layout just after
	 */
    @Override
	public void setContentView(int layoutResID) {
		super.setContentView(layoutResID);

        mContentView = getWindow().getDecorView().findViewById(android.R.id.content);
    }

	public View getContentView() {
		return mContentView;
	}

	/**
	 * Each activity instance needs its own HistoryTraversal object
	 * @param historyTraversalFactory
	 */
	@Inject
	void setNewHistoryTraversal(HistoryTraversalFactory historyTraversalFactory) {
		// Ensure we don't end up overwriting the initialised class
		if (historyTraversal==null) {
			this.historyTraversal = historyTraversalFactory.createHistoryTraversal(integrateWithHistoryManagerInitialValue);
		}
	}

	@Inject
	void setSwordDocumentFacade(SwordDocumentFacade swordDocumentFacade) {
		this.swordDocumentFacade = swordDocumentFacade;
	}

	protected SwordDocumentFacade getSwordDocumentFacade() {
		return swordDocumentFacade;
	}
}

File: app/src/main/java/net/bible/android/view/activity/base/ActivityBase.kt
/*
 * Copyright (c) 2020-2022 Martin Denham, Tuomas Airaksinen and the AndBible contributors.
 *
 * This file is part of AndBible: Bible Study (http://github.com/AndBible/and-bible).
 *
 * AndBible is free software: you can redistribute it and/or modify it under the
 * terms of the GNU General Public License as published by the Free Software Foundation,
 * either version 3 of the License, or (at your option) any later version.
 *
 * AndBible is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with AndBible.
 * If not, see http://www.gnu.org/licenses/.
 */

package net.bible.android.view.activity.base

import android.annotation.SuppressLint
import android.app.Activity
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.Bundle
import android.util.Log
import android.view.KeyEvent
import android.view.View
import android.view.WindowManager
import android.view.inputmethod.InputMethodManager
import androidx.activity.result.ActivityResult
import androidx.activity.result.IntentSenderRequest
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import net.bible.android.view.activity.StartupActivity
import net.bible.android.view.activity.comingFromStartupActivity
import net.bible.android.view.activity.discrete.CalculatorActivity
import net.bible.android.view.util.UiUtils.setActionBarColor
import net.bible.android.view.util.locale.LocaleHelper
import net.bible.service.common.CommonUtils
import net.bible.service.device.ScreenSettings
import net.bible.service.history.HistoryTraversal
import net.bible.service.history.HistoryTraversalFactory
import javax.inject.Inject

var firstTime = true

/** Base class for activities
 *
 * @author Martin Denham [mjdenham at gmail dot com]
 */
abstract class ActivityBase : AppCompatActivity(), AndBibleActivity {
    private var isScreenOn = true

    // some screens are highly customised and the theme looks odd if it changes
    open val allowThemeChange = true
    open val integrateWithHistoryManager: Boolean = false

    protected lateinit var historyTraversal: HistoryTraversal

    open val doNotInitializeApp = false

    private var doNotMarkPaused = false
    private var wasPaused = false
    private var returningFromCalculator = false

    /** Called when the activity is first created.  */
    @SuppressLint("MissingSuperCall")
    public override fun onCreate(savedInstanceState: Bundle?) {
        CurrentActivityHolder.activate(this)

        if(!doNotInitializeApp) {
            CommonUtils.initializeApp()
        }

        if (allowThemeChange) {
            applyTheme()
        }

        super.onCreate(savedInstanceState)

        if(!doNotInitializeApp) {
            if(CommonUtils.showCalculator) {
                window.addFlags(WindowManager.LayoutParams.FLAG_SECURE)
            }
            refreshScreenKeepOn()
        }

        Log.i(TAG, "onCreate")

        // if locale is overridden then have to force title to be translated here
        LocaleHelper.translateTitle(this)
        setActionBarColor(supportActionBar)

        wasPaused = false
        Log.i(TAG, "onCreate: loading state: $savedInstanceState")
        if(savedInstanceState != null) {
            doNotMarkPaused = savedInstanceState.getBoolean("doNotMarkPaused", false)
            wasPaused = savedInstanceState.getBoolean("wasPaused", false)
            returningFromCalculator = savedInstanceState.getBoolean("returningFromCalculator", false)
        }
        fixNightMode()
    }

    open fun fixNightMode() {
        // First launched activity is not having proper night mode if we are using manual mode.
        // This hack fixes it.
        if(firstTime && allowThemeChange && !doNotInitializeApp) {
            firstTime = false
            lifecycleScope.launch {
                delay(250)
                recreate()
            }
            return
        }
    }

    fun applyTheme() {
        val newNightMode = if (ScreenSettings.nightMode) {
            AppCompatDelegate.MODE_NIGHT_YES
        } else {
            AppCompatDelegate.MODE_NIGHT_NO
        }
        Log.i(TAG, "applyTheme: nightMode = ${ScreenSettings.nightMode}")
        AppCompatDelegate.setDefaultNightMode(newNightMode)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            if (!ScreenSettings.nightMode) {
                val uiFlags = window.decorView.systemUiVisibility or View.SYSTEM_UI_FLAG_LIGHT_NAVIGATION_BAR
                window.decorView.systemUiVisibility = uiFlags
            }
        }
    }

    protected fun buildActivityComponent() = CommonUtils.buildActivityComponent()

    override fun onSaveInstanceState(outState: Bundle) {
        outState.putBoolean("doNotMarkPaused", doNotMarkPaused)
        outState.putBoolean("wasPaused", wasPaused)
        outState.putBoolean("returningFromCalculator", returningFromCalculator)
        Log.i(TAG, "Saving saved state from $outState")

        super.onSaveInstanceState(outState)
    }

    override fun startActivity(intent: Intent) {
        if(integrateWithHistoryManager) {
            historyTraversal.beforeStartActivity()
        }

        super.startActivity(intent)
    }

    override fun startActivityForResult(intent: Intent, requestCode: Int) {
        if(integrateWithHistoryManager) {
            historyTraversal.beforeStartActivity()
        }

        super.startActivityForResult(intent, requestCode)
    }

    /**
     * Override locale.  If user has selected a different ui language to the devices default language
     */
    override fun attachBaseContext(newBase: Context) {
        super.attachBaseContext(LocaleHelper.onAttach(newBase))
    }

    /**	This will be called automatically for you on 2.0 or later
     */
    override fun onBackPressed() {
        if (!historyTraversal.goBack()) {
            super.onBackPressed()
        }
    }

    /** called by Android 2.0 +
     */
    override fun onKeyLongPress(keyCode: Int, event: KeyEvent): Boolean {
        // ignore long press on search because it causes errors
        if (keyCode == KeyEvent.KEYCODE_SEARCH) {
            // ignore
            return true
        }

        //TODO make Long press back - currently the History screen does not show the correct screen after item selection if not called from main window
        return if (keyCode == KeyEvent.KEYCODE_BACK) {
            // ignore
            true
        } else super.onKeyLongPress(keyCode, event)

    }

    override var isIntegrateWithHistoryManager: Boolean
        get() = historyTraversal.isIntegrateWithHistoryManager
        set(value) {
            historyTraversal.isIntegrateWithHistoryManager = value
        }

    /** allow activity to enhance intent to correctly restore state  */
    override val intentForHistoryList: Intent get() = intent

    fun showErrorMsg(msgResId: Int) {
        Dialogs.showErrorMsg(msgResId)
    }

    protected fun returnErrorToPreviousScreen() {
        // just pass control back to the previous screen
        val resultIntent = Intent(this, this.javaClass)
        setResult(Activity.RESULT_CANCELED, resultIntent)
        finish()
    }

    protected fun returnToPreviousScreen() {
        // just pass control back to the previous screen
        val resultIntent = Intent(this, this.javaClass)
        setResult(Activity.RESULT_OK, resultIntent)
        finish()
    }

    protected fun returnToTop() {
        // just pass control back to the previous screen
        val resultIntent = Intent(this, this.javaClass)
        setResult(RESULT_RETURN_TO_TOP, resultIntent)
        finish()
    }

    override fun onResume() {
        CurrentActivityHolder.activate(this)
        super.onResume()
        Log.i(TAG, "onResume wasPaused:$wasPaused returningFromCalculator:$returningFromCalculator")
        val fromStartupActivity = comingFromStartupActivity
        comingFromStartupActivity = false
        if (
            this !is CalculatorActivity
            && !fromStartupActivity
            && this !is StartupActivity
            && CommonUtils.showCalculator
            && wasPaused
            && !returningFromCalculator
        ) {
            val handlerIntent = Intent(this@ActivityBase, CalculatorActivity::class.java)
            startActivityForResult(handlerIntent, CALCULATOR_REQUEST)
            returningFromCalculator = true
        } else {
            returningFromCalculator = false
        }
        wasPaused = false

        //allow action to be called on screen being turned on
        if (!isScreenOn && ScreenSettings.isScreenOn) {
            onScreenTurnedOn()
        }
    }

    override fun startActivity(intent: Intent?, options: Bundle?) {
        doNotMarkPaused = true
        super.startActivity(intent, options)
    }

    override fun startActivityForResult(intent: Intent, requestCode: Int, options: Bundle?) {
        doNotMarkPaused = true
        super.startActivityForResult(intent, requestCode, options)
    }

    override fun onPause() {
        super.onPause()
        if(!doNotMarkPaused) {
            wasPaused = true
        }
        doNotMarkPaused = false
        Log.i(TAG, "onPause: $this")
        if (isScreenOn && !ScreenSettings.isScreenOn) {
            onScreenTurnedOff()
        }
        closeKeyboard()
    }


    fun closeKeyboard() {
        try {
            val inputMethodManager: InputMethodManager = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
            inputMethodManager.hideSoftInputFromWindow(currentFocus!!.windowToken, 0)
        } catch (e: Exception) {
            Log.e(TAG, "closeKeyboard: $e")
        }
    }

    protected open fun onScreenTurnedOff() {
        Log.i(TAG, "Window turned off")
        isScreenOn = false
    }

    protected open fun onScreenTurnedOn() {
        Log.i(TAG, "Window turned on")
        isScreenOn = true
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.i(TAG, "onDestroy")
    }

    override fun onRestart() {
        super.onRestart()
        if(!doNotInitializeApp) {
            refreshScreenKeepOn()
        }
        Log.i(TAG, "onRestart")
    }

    override fun onStart() {
        super.onStart()
        Log.i(TAG, "onStart")
        CommonUtils.onyxSupport?.setupOnyxFast()
    }

    override fun onNewIntent(intent: Intent?) {
        super.onNewIntent(intent)
        Log.i(TAG, "onNewIntent $this ${intent?.action}")
    }

    override fun onStop() {
        super.onStop()
        Log.i(TAG, "onStop")
        CurrentActivityHolder.deactivate(this)
    }

    /**
     * Each activity instance needs its own HistoryTraversal object
     * @param historyTraversalFactory
     */
    @Inject
    internal fun setNewHistoryTraversal(historyTraversalFactory: HistoryTraversalFactory) {
        // Ensure we don't end up overwriting the initialised class
        if (!::historyTraversal.isInitialized) {
            this.historyTraversal = historyTraversalFactory.createHistoryTraversal(integrateWithHistoryManager)
        }
    }

    private var currentCode : Int = 0
    private var resultByCode = mutableMapOf<Int, CompletableDeferred<ActivityResult>>()

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        Log.i(TAG, "onActivityResult: requestCode = $requestCode, resultCode = $resultCode, data is${if (data != null) " not" else ""} null")
        if(requestCode == CALCULATOR_REQUEST) {
            if(resultCode == RESULT_CANCELED) {
                finishAffinity()
            }
            return
        }
        resultByCode[requestCode - ASYNC_REQUEST_CODE_START]?.let {
            it.complete(ActivityResult(resultCode, data))
            resultByCode.remove(requestCode - ASYNC_REQUEST_CODE_START)
        } ?: run {
            super.onActivityResult(requestCode, resultCode, data)
        }
    }

    suspend fun awaitIntent(intent: Intent) : ActivityResult
    {
        val activityResult = CompletableDeferred<ActivityResult>()
        val resultCode = currentCode++
        resultByCode[resultCode] = activityResult
        startActivityForResult(intent, resultCode + ASYNC_REQUEST_CODE_START)
        return activityResult.await()
    }

    val preferences get() = CommonUtils.settings

    private var deferredActivityResult = CompletableDeferred<ActivityResult>()
    private val deferredActivityResultMutex = Mutex()

    private val intentSenderLauncher =
        registerForActivityResult(ActivityResultContracts.StartIntentSenderForResult()) {
            deferredActivityResult.complete(it)
        }
    suspend fun awaitPendingIntent(pendingIntent: PendingIntent): ActivityResult = deferredActivityResultMutex.withLock {
        val defer = CompletableDeferred<ActivityResult>()
        deferredActivityResult = defer
        intentSenderLauncher.launch(
            IntentSenderRequest.Builder(pendingIntent.intentSender).build()
        )
        return defer.await()
    }


    private fun refreshScreenKeepOn() {
        val keepOn = preferences.getBoolean(SCREEN_KEEP_ON_PREF, false)
        if (keepOn) {
            window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        } else {
            window.clearFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)
        }
    }

    open fun freeze() {}

    open fun unFreeze() {}

    val TAG get() = "Base-${this::class.java.simpleName}"

    companion object {
        private const val SCREEN_KEEP_ON_PREF = "screen_keep_on_pref"

        // standard request code for startActivityForResult
        const val STD_REQUEST_CODE = 1
        const val CALCULATOR_REQUEST = 6000
        const val ASYNC_REQUEST_CODE_START = 1900

        // Special result that requests all activities to exit until the main/top Activity is reached
        const val RESULT_RETURN_TO_TOP = 900

    }
}
